exatomic/nwchem/output.py

```diff
@@ -257,6 +264,167 @@ def parse_basis_set_order(self):
+        _renat = "Atom information"
+
+        found = self.find(_remeth)
+        fnat = self.find(_renat)
+        if not found and not fnat:
+            return
+        # get atom information
+        start = fnat[0][0]+3
+        stop = start
+        while '----' not in self[stop]: stop += 1
+        # we assume that there is only one instance of where _renat is found
+        columns = ['symbol', 'atom', 'x', 'y', 'z', 'mass']
+        atom = self.pandas_dataframe(start, stop, columns)
+        atom['atom'] -= 1
+        nat = len(atom)
+        # find bounds where the calculated frequencies are
+        start = found[0][0]
+        stop = found[1][0]
+        # get the data
+        found = self.find(_refreq, start=start, stop=stop)
+        dfs = []
+        fdx = 0
+        # get frequencies
+        for lno, ln in found:
+            # get the frequency values
+            tmp = ln.split()[1:]
+            freq = np.asarray([float(i) for i in tmp])
+            ## TODO: here we remove all negative frequencies
+            ##       need to find out if this is ok to do
+            # set start and end points for the calculated normal modes
+            staf = lno+start+1
+            stof = lno+start+nat*3+2
+            nm = self.pandas_dataframe(staf, stof, ncol=len(freq)).reset_index(drop=True)
+            # generate boolean array that shows False for negative frequencies
+            neg = [False if f < 0 else True for f in freq]
+            # remove negative frequencies
+            nm.drop(columns=[idx for idx, val in enumerate(neg) if not val], inplace=True)
+            freq = freq[neg]
+            # get normal modes in the x, y, z directions
+            nm = nm.stack().values
+            nfreq = len(freq)
+            dx = nm[::3]
+            dy = nm[1::3]
+            dz = nm[2::3]
+            # assemble dataframe
+            symbol = np.tile(atom['symbol'], nfreq)
+            adx = np.tile(atom['atom'], nfreq)
+            freq = np.repeat(freq, nat)
+            freqdx = np.repeat([i for i in range(fdx, fdx + nfreq)], nfreq)
+            frames = np.repeat([0], nfreq*nat)
+            fdx += nfreq
+            stacked = pd.DataFrame.from_dict({'symbol': symbol, 'atom': adx, 'dx': dx, 'dy': dy,
+                                              'dz': dz, 'freq': freq, 'freqdx': freqdx,
+                                              'frames': frames})
+            dfs.append(stacked)
+        frequency = pd.concat(dfs).reset_index(drop=True)
+        self.frequency = frequency
+
+    def parse_gradient(self):
+        """
+        Parse :class:`exatomic.core.gradient.Gradient` dataframe.
+        """
+        _regrad = "DFT ENERGY GRADIENTS"
+
+        found = self.find(_regrad)
+        if not found:
+            return
+        found = self.find(_regrad, keys_only=True)
+        # find start and stop points
+        starts = np.array(found) + 4
+        stop = starts[0]
+        while '----' not in self[stop]: stop += 1
```

Comment:
I agree I feel that it may be more useful to add something in the while loop that throws an error saying that it reached the EOF and did not find the pattern, because that probably indicates an issue with the file being read. Be it, that the program changed the characters encapsulating the data or that the data on the file is incomplete.

Reply:
Thinking on it a bit how do you think this would work
```
while '----' not in self[stop]:
    stop += 1
    if stop >= len(self):
        raise EOFError("Reached EOF while trying to parse_gradients. 
                                   Make sure the file is complete.")
```
I'm basically trying to go for detailed errors that give the user a reason why this failed. Also, if there is an issue with the code this will give an alert and if there is nothing wrong with the file we can look into why this would give that error.
Don't know if it might be a good idea to look into doing this with some of the other parsing codes that use a similar structure with the while loops.

Reply:
I think this might be overkill for this case maybe? An `IndexError` will be raised if `self[stop]` can't be accessed right? Maybe just leave it as is?

Reply:
Probably, I just always try to make helpful error messages because half the time you get an error and it basically says "you made a mistake, I really don't care so go away".

Reply:
This is another good example of a place where logging could be a big assist.  You can run your while loop inside a try except for the IndexError directly, and rather than terminating the process by raising an explicit exception, log the pertinent information as part of handling the exception, and carry on with business.

```python
try:
    while '----' not in self[stop]:
        # do things
except IndexError as e:
    self.log.error('parsing gradients failed. check file for completeness')
    # cleanup or raise custom exception explicitly; optionally passing e 
```